mypy/checkexpr.py

```diff
@@ -682,30 +686,67 @@ def check_typeddict_call_with_dict(
-            ]
-            actual_keys = kwargs.keys()
+        """Check TypedDict constructor of format Point(x=1, y=2)"""
+        # Infer types of item values and expand unpack operators
+        items: Dict[str, Tuple[Expression, Type]] = {}
+        sure_keys: List[str] = []
+        maybe_keys: List[str] = []  # Will contain non-required items of unpacked TypedDicts
+        for key, value_expr in kwargs:
+            if key is not None:
+                # Regular key and value
+                value_type = self.accept(value_expr, callee.items.get(key))
+                items[key] = (value_expr, value_type)
+                sure_keys.append(key)
+            else:
+                # Unpack operator (**) was used; unpack all items of the type of this expression into items list
+                value_type = self.accept(value_expr, callee)
+                proper_type = get_proper_type(value_type)
+                if isinstance(proper_type, TypedDictType):
+                    for nested_key, nested_value_type in proper_type.items.items():
+                        items[nested_key] = (value_expr, nested_value_type)
+                        if nested_key in proper_type.required_keys:
+                            sure_keys.append(nested_key)
+                        else:
+                            maybe_keys.append(nested_key)
+                else:
+                    # Fail when trying to unpack anything but TypedDict
+                    assert not self.chk.check_subtype(
+                        subtype=value_type,
+                        supertype=self.chk.named_type("typing._TypedDict"),
+                        context=value_expr,
+                        msg=message_registry.INCOMPATIBLE_TYPES,
+                        subtype_label="unpacked expression has type",
+                        supertype_label="expected",
+                        code=codes.TYPEDDICT_ITEM,
+                    )
+
+        if not (
+            callee.required_keys
+            <= set(sure_keys)
+            <= set(sure_keys + maybe_keys)
+            <= set(callee.items.keys())
+        ):
             self.msg.unexpected_typeddict_keys(
-                callee, expected_keys=expected_keys, actual_keys=list(actual_keys), context=context
+                callee, actual_sure_keys=sure_keys, actual_maybe_keys=maybe_keys, context=context
             )
             return AnyType(TypeOfAny.from_error)
 
+        # Check item value types
         for (item_name, item_expected_type) in callee.items.items():
-            if item_name in kwargs:
-                item_value = kwargs[item_name]
-                self.chk.check_simple_assignment(
-                    lvalue_type=item_expected_type,
-                    rvalue=item_value,
-                    context=item_value,
+            if item_name in items:
+                item_value_expr, item_actual_type = items[item_name]
+                self.chk.check_subtype(
```

Comment:
I'm calling `check_subtype` instead of `check_simple_assignment` because that allows consistent handling of both immediate and unpacked items. Let me know if `check_simple_assignment` does any important additional checks that might be important